library/java/io/envoyproxy/envoymobile/engine/AndroidProxyMonitor.java

```diff
@@ -45,15 +46,41 @@ private void registerReceiver(Context context) {
 
   @Override
   public void onReceive(Context context, Intent intent) {
-    handleProxyChange();
+    handleProxyChange(intent);
   }
 
-  private void handleProxyChange() {
-    ProxyInfo info = connectivityManager.getDefaultProxy();
+  private void handleProxyChange(final Intent intent) {
+    ProxyInfo info = this.extractProxyInfo(intent);
+
     if (info == null) {
       envoyEngine.setProxySettings("", 0);
     } else {
       envoyEngine.setProxySettings(info.getHost(), info.getPort());
     }
   }
+
+  private ProxyInfo extractProxyInfo(final Intent intent) {
+    ProxyInfo info = connectivityManager.getDefaultProxy();
+
+    // If proxy is configured using the PAC file use the
+    // Android's injected localhost HTTP proxy.
+    //
+    // Android's injected localhost proxy can be accessed using a proxy host
+    // equal to `localhost` and a proxy port retrieved from intent's 'extras'.
+    // We cannot take a proxy port from the ProxyInfo object that's exposed by
+    // the connectivity manager as it's always equal to -1 for cases when PAC
+    // proxy is configured.
+    //
+    // See https://github.com/envoyproxy/envoy-mobile/issues/2531 for more details.
+    if (info.getPacFileUrl() != null && info.getPacFileUrl() != Uri.EMPTY) {
```

Comment:
Do you have to check if `info` is `null` too to avoid an NPE?

Reply:
Talked with @murki - he suggested we leave it, if anything to make sure that this not crash on any Android version (including the older ones)

Reply:
I mean in addition to your current checks, also do `info != null`

```suggestion
    if (info != null && info.getPacFileUrl() != null && info.getPacFileUrl() != Uri.EMPTY) {
```